api/src/main/java/org/asynchttpclient/request/body/generator/FeedableBodyGenerator.java

```diff
@@ -74,51 +75,47 @@ public long getContentLength() {
 
         @Override
         public long read(final ByteBuffer buffer) throws IOException {
-            BodyPart nextPart = queue.peek();
-            if (nextPart == null) {
-                // Nothing in the queue
-                switch (state) {
+            switch (state) {
                 case ONGOING:
-                    return 0;
-                case CLOSING:
-                    buffer.put(ZERO);
-                    buffer.put(END_PADDING);
-                    buffer.put(END_PADDING);
-                    state = PushBodyState.FINISHED;
-                    return buffer.position();
+                    return readNextPart(buffer);
                 case FINISHED:
                     return -1;
-                }
-            }
-            if (nextPart.buffer.remaining() == 0) {
-                // skip empty buffers
-                // if we return 0 here it would suspend the stream - we don't want that
-                queue.remove();
-                if (nextPart.isLast) {
-                    state = writeChunkBoundaries ? PushBodyState.CLOSING : PushBodyState.FINISHED;
-                }
-                return read(buffer);
+                default:
+                    throw new IllegalStateException("Illegal process state.");
             }
-            int capacity = buffer.remaining() - 10; // be safe (we'll have to add size, ending, etc.)
-            int size = Math.min(nextPart.buffer.remaining(), capacity);
-            if (size != 0) {
-                if (writeChunkBoundaries) {
-                    buffer.put(Integer.toHexString(size).getBytes(US_ASCII));
-                    buffer.put(END_PADDING);
-                }
-                for (int i = 0; i < size; i++) {
-                    buffer.put(nextPart.buffer.get());
+        }
+
+        private long readNextPart(ByteBuffer buffer) throws IOException {
+            int reads = 0;
+            while (buffer.hasRemaining() && state != PushBodyState.FINISHED) {
+                BodyPart nextPart = queue.peek();
+                if (nextPart == null) {
+                    // Nothing in the queue. suspend stream if nothing was read. (reads == 0)
+                    return reads;
+                } else if (!nextPart.buffer.hasRemaining() && !nextPart.isLast) {
+                    // skip empty buffers
+                    // if we return 0 here it would suspend the stream - we don't want that
+                    queue.remove();
+                } else {
+                    readBodyPart(buffer, nextPart);
+                    reads++;
                 }
-                if (writeChunkBoundaries)
-                    buffer.put(END_PADDING);
             }
-            if (!nextPart.buffer.hasRemaining()) {
-                if (nextPart.isLast) {
-                    state = writeChunkBoundaries ? PushBodyState.CLOSING : PushBodyState.FINISHED;
+            return reads;
```

Comment:
I guess it could just return the difference between the buffer current and original positions.